File: algorithm/337. House Robber III/java/HouseRobberIII.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// use a size 2 array to record cur and prev as did in House Robber I and II.
// not sure if this is optimal.
// notice if trying to re-write it in C or C++ or other language that requires memory management, remember to drop unused memory
class Solution {
    public int rob(TreeNode root) {
        int[] result = recur(root);
        return result[0];
    }
    
    private int[] recur(TreeNode root){
        if(root==null) return new int[2];
        int[] left = recur(root.left);
        int[] right = recur(root.right);
        int result = 0;
        result = Math.max(left[0]+right[0], root.val+left[1]+right[1]);
        left[1] = left[0]+right[0];
        left[0] = result;
        return left;
    }
}

File: algorithm/337. House Robber III/cpp/HouseRobberIII.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
// use std::pair would make the code look cleaner
class Solution {
public:
    int rob(TreeNode* root) {
        vector<int> temp = recur(root);
        int result = temp[0];
        // release memory
        vector<int>().swap(temp);
        return result;
    }
    
private:
    vector<int> recur(TreeNode* root){
        if(!root) return vector<int>(2,0);
        vector<int> left = recur(root->left);
        vector<int> right = recur(root->right);
        int curMax = max(left[0]+right[0], root->val+left[1]+right[1]);
        left[1] = left[0]+right[0];
        left[0] = curMax;
        // release memory
        vector<int>().swap(right);
        return left;
    }
};
